Editor/AGS.Editor/Components/RoomsComponent.cs

```diff
@@ -1545,20 +1543,38 @@ void IRoomController.SetBackground(int background, Bitmap bmp)
                 throw new ArgumentNullException(nameof(bmp));
             }
 
-            _loadedRoom.Width = bmp.Width;
-            _loadedRoom.Height = bmp.Height;
-            // TODO Replace with bitmap saving to disk with C# when room is open format
-            _nativeProxy.ImportBackground(
-                _loadedRoom, background, bmp, _agsEditor.Settings.RemapPalettizedBackgrounds, sharePalette: false);
+            Bitmap newBmp = (Bitmap)bmp.Clone();
+            _loadedRoom.Width = newBmp.Width;
+            _loadedRoom.Height = newBmp.Height;
+            _loadedRoom.ColorDepth = newBmp.GetColorDepth();
+
+            if (_loadedRoom.ColorDepth == 8)
+            {
+                ColorPalette palette = newBmp.Palette;
+                foreach (var color in _agsEditor.CurrentGame.Palette.Where(p => p.ColourType == PaletteColourType.Locked))
+                {
+                    palette.Entries[color.Index] = color.Colour;
+                }
+                newBmp.Palette = palette;
+
+                newBmp.SetGlobalPaletteFromPalette();
+            }
 
             if (background >= _backgroundCache.Count)
             {
-                _backgroundCache.Add((Bitmap)bmp.Clone());
+                _backgroundCache.Add(newBmp);
+                _loadedRoom.BackgroundCount++;
             }
             else
             {
                 _backgroundCache[background]?.Dispose();
-                _backgroundCache[background] = (Bitmap)bmp.Clone();
+                _backgroundCache[background] = newBmp;
+            }
+
+            // If size or resolution has changed, reset masks
+            if (newBmp.Width != _loadedRoom.Width || newBmp.Height != _loadedRoom.Height)

```

Comment:
Ah... so this is why masks are never resetted. This check is always false because _loadedRoom.Width/Height have already been changed to that same size.

But even fixing that issue, ClearMaskCache() disposes of the masks without recreating them, so we later get null pointer exceptions.